include/oneapi/dpl/pstl/zip_view_impl.h

```diff
@@ -0,0 +1,401 @@
+        operator-(const sentinel& y, const iterator<OtherConst>& x) {
+            return -(x - y);
+        }
+
+    private:
+        friend class zip_view;
+
+        using end_type = std::conditional_t<!Const, tuple_type<std::ranges::sentinel_t<Views>...>,
+                                                    tuple_type<std::ranges::sentinel_t<const Views>...>>;
+
+        end_type end_;
+    }; // class sentinel
+
+    constexpr auto begin() requires (std::ranges::range<Views> && ...)
+    {
+        auto __tr = [](auto... __args) { return iterator<false>(__args...);};
+        return apply_to_tuple(__tr, std::ranges::begin, views_);
+    }
+
+    constexpr auto begin() const requires ( std::ranges::range<const Views> && ... )
+    {
+        auto __tr = [](auto... __args) { return iterator<true>(__args...);};
+        return apply_to_tuple(__tr, std::ranges::begin, views_);
+    }
+
+    constexpr auto end() requires (std::ranges::range<Views> && ...)
+    {
+        if constexpr (!zip_is_common<Views...>)
+        {
+            auto __tr = [](auto... __args) { return sentinel<false>(__args...);};
+            return apply_to_tuple(__tr, std::ranges::end, views_);
+        }
+        else if constexpr ((std::ranges::random_access_range<Views> && ...))
+        {
+            auto it = begin();
+            it += size();
+            return it;
+        }
+        else
+        {
+            auto __tr = [](auto... __args) { return iterator<false>(__args...);};
+            return apply_to_tuple(__tr, std::ranges::end, views_);
+        }
+    }
+
+    constexpr auto end() const requires (std::ranges::range<const Views> && ...)
+    {
+        if constexpr (!zip_is_common<Views...>)
+        {
+            auto __tr = [](auto... __args) { return sentinel<true>(__args...);};
+            return apply_to_tuple(__tr, std::ranges::end, views_);
+        } 
+        else if constexpr ((std::ranges::random_access_range<Views> && ...))
+        {
+            auto it = begin();
+            it += size();
+            return it;
+        }
+        else
+        {
+            auto __tr = [](auto... __args) { return iterator<true>(__args...);};
+            return apply_to_tuple(__tr, std::ranges::end, views_);
+        }
+    }
+
+    constexpr auto size() requires (std::ranges::sized_range<Views> && ...)
+    {
+        auto __tr = [](auto... __args) { 
+            using CT = std::make_unsigned_t<std::common_type_t<decltype(__args)...>>;
+            return std::ranges::min({CT(__args)...});
+            };
+
+        return apply_to_tuple(__tr, std::ranges::size, views_);
+    }
+
+    constexpr auto size() const requires (std::ranges::sized_range<Views> && ...)
+    {
+        return const_cast<zip_view*>(this)->size();
+    }
+private:
+    tuple_type<Views...> views_;
+}; // class zip_view
+
+template <typename... Rs>
+zip_view(Rs&&...) -> zip_view<std::views::all_t<Rs>...>;
+
```

Comment:
Do we want to add a specialization for `enable_borrowed_range`? 
See here for reference: https://en.cppreference.com/w/cpp/ranges/zip_view#:~:text=template%3C%20class,view%20satisfies%20it.

Reply:
Rather we don't, due to it is "since C++23"...